Sources/xcodeproj/Utils/PBXBatchUpdater.swift

```diff
@@ -0,0 +1,84 @@
+import Foundation
+import PathKit
+
+// This is a helper class for quickly adding a large number of files.
+// It is forbidden to add a file to a group one by one using the PBXGroup method addFile(...) while you are working with this class.
+public final class PBXBatchUpdater {
+    
+    private let projectObjects: PBXObjects
+    private let sourceRoot: Path
+    private var references: [Path: PBXObjectReference]? = nil
+    
+    init(projectObjects: PBXObjects, sourceRoot: Path) {
+        self.projectObjects = projectObjects
+        self.sourceRoot = sourceRoot
+    }
+    
+    /// Adds file at the give path to the project or returns existing file and its reference.
+    ///
+    /// - Parameters:
+    ///   - filePath: path to the file.
+    ///   - sourceTree: file sourceTree, default is `.group`
+    ///   - sourceRoot: path to project's source root.
+    /// - Returns: new or existing file and its reference.
+    @discardableResult
+    public func addFile(
+        to group: PBXGroup,
+        at filePath: Path,
+        sourceTree: PBXSourceTree = .group
+        ) throws -> PBXFileReference {
+        
+        guard filePath.exists else {
+            throw XcodeprojEditingError.unexistingFile(filePath)
+        }
+        let groupPath = try group.fullPath(sourceRoot: sourceRoot)
+        
+        // Lazy initialization
+        let objectReferences: [Path: PBXObjectReference]
+        if let references = self.references {
+            objectReferences = references
+        } else {
+            objectReferences = Dictionary(uniqueKeysWithValues:
+                try projectObjects.fileReferences.compactMap({
+                    guard let fullPath = try $0.value.fullPath(sourceRoot: sourceRoot) else { return nil }
+                    return (fullPath, $0.key)
+                })
+            )
+            references = objectReferences
+        }
+        
+        if let existingObjectReference = objectReferences[filePath],
+            let existingFileReference = projectObjects.fileReferences[existingObjectReference] {
+            if !group.childrenReferences.contains(existingObjectReference) {
+                existingFileReference.path = groupPath.flatMap({ filePath.relative(to: $0) })?.string
```

Comment:
I like the idea of the batch updater lazily creating a variable that allows us to get the object file reference given a path. However, I think this is not quite exactly the way Xcode adds files. Let's say we have the following scenario:

- Group: 
  - Name: `Sources`
  - Path: `/path/to/project/Sources`

- File we want to add:
  - Name: `SearchViewController.swift`
  - Path: `/path/to/project/Sources/Features/SearchViewController.swift`

If I call `addFile` passing the sources group and the absolute path to the `SearchViewController.swift` file. I'll end up getting the following on the Xcode project:

- Sources
   - SearchViewController.swift

When what I'd expect is the following:

- Sources
  - Features
    - Search
      - SearchViewController

Reply:
I'd adjust the logic here a little bit to be like this *(continuing with the example from the comment above)*.

I'd take the absolute path of the file that I'd like to add, in this case `/path/to/project/Sources/Features/SearchViewController.swift` and I'dstart removing components from the right until we find a group that is part of the project. We'd find that `/path/to/project/Sources/` is already part of the project so we take that group and start creating the children down from that group. It'd result in the following:

- A `Features` group child of `Sources`.
- A `Search` group child of `Features`.
- A `SearchViewController.swift` file reference child of `Search`.